metal/src/utils/device_wrapper.mm

```diff
@@ -0,0 +1,142 @@
+/* VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+// Local include(s).
+#include "vecmem/utils/metal/device_wrapper.hpp"
+#include "get_device.h"
+#include "opaque_device.h"
+#include "vecmem/utils/debug.hpp"
+
+// System import(s).
+#include <Metal/Metal.h>
+
+// System include(s).
+#include <cassert>
+
+namespace vecmem::metal {
+
+device_wrapper::device_wrapper()
+    : m_device(nullptr),
+      m_managedDevice(std::make_unique<details::opaque_device>()) {
+
+    // Create a (default) new Metal device.
+    m_managedDevice->m_device = MTLCreateSystemDefaultDevice();
+    if (m_managedDevice->m_device != nil) {
+
+        // Grab a bare pointer to the object.
+        m_device = (__bridge void*)m_managedDevice->m_device;
+
+        // Tell the user what happened.
+        VECMEM_DEBUG_MSG(
+            1, "Created an \"owning wrapper\" around device: %s",
+            [[details::get_device(*this) name]
+                cStringUsingEncoding:[NSString defaultCStringEncoding]]);
+
+    } else {
+        // Tell the user what happened.
+        VECMEM_DEBUG_MSG(1, "No Metal device was found");
+    }
+}
+
+device_wrapper::device_wrapper(void* device)
+    : m_device(device), m_managedDevice() {
+
+    assert(m_device != nullptr);
+
+    // Tell the user what happened.
+    VECMEM_DEBUG_MSG(
+        3, "Created a \"view wrapper\" around device: %s",
+        [[details::get_device(*this) name]
+            cStringUsingEncoding:[NSString defaultCStringEncoding]]);
+}
+
+device_wrapper::device_wrapper(const device_wrapper& parent)
+    : m_device(nullptr), m_managedDevice() {
+
+    // Check whether the parent owns its own device or not.
+    if (parent.m_managedDevice) {
+        // If so, make a copy of it, and own that copy in this object as well.
+        // This makes sure that reference counts to the MTLDevice would be
+        // managed properly.
+        m_managedDevice =
+            std::make_unique<details::opaque_device>(*(parent.m_managedDevice));
+        m_device = (__bridge void*)m_managedDevice->m_device;
```

Comment:
What does `__bridge` do here?

Reply:
I'm on **very** shaky ground here. 🤔 I was trying to find out from the internets how to do the equivalent of what's happening in `vecmem::sycl::queue_wrapper`. Where a `void*` pointer is used to point at a `cl::sycl::queue` object in memory.

The compiler was happy to let me convert an `id<MTLDevice>` "object" to a `void*` pointer, but I was not sure how to "safely" do the reverse. Since the `reinterpret_cast` that we do in `vecmem::sycl::queue_wrapper` is not really an option here. (I think...)

After some looking around, I found: https://clang.llvm.org/docs/AutomaticReferenceCounting.html#bridged-casts Tried to use it, and the code compiled without warnings, and ran successfully. 🤔 So I went with this formalism. However it's absolutely true that this code might be doing something silly at this point.